src/pipe_windows.rs

```diff
@@ -0,0 +1,204 @@
+        );
+    }
+}
+
+pub(crate) fn wake_socket(pipe: RawSocket) {
+    unsafe {
+        send(pipe as SOCKET, b"X" as *const _ as *const _, 1, 0);
+    }
+}
+
+/// Registers a write to a self-pipe whenever there's the signal.
+///
+/// In this case, the pipe is taken as the `RawHandle`. It is still the caller's responsibility to
+/// close it.
+pub fn register_handle_raw(signal: c_int, pipe: RawHandle) -> Result<SigId, Error> {
+    struct SendHandle(RawHandle);
+    unsafe impl Send for SendHandle {}
+    unsafe impl Sync for SendHandle {}
+    let pipe = SendHandle(pipe);
+    unsafe { ::register(signal, move || wake_handle(pipe.0)) }
+}
+
+/// Registers a write to a self-pipe whenever there's the signal.
+///
+/// In this case, the pipe is taken as the `RawSocket`. It is still the caller's responsibility to
+/// close it.
+pub fn register_socket_raw(signal: c_int, pipe: RawSocket) -> Result<SigId, Error> {
+    struct SendSocket(RawSocket);
+    unsafe impl Send for SendSocket {}
+    unsafe impl Sync for SendSocket {}
+    let pipe = SendSocket(pipe);
+    unsafe { ::register(signal, move || wake_socket(pipe.0)) }
+}
+
+/// Registers a write to a self-pipe whenever there's the signal.
+///
+/// The ownership of pipe is taken and will be closed whenever the created action is unregistered.
+///
+/// Note that if you want to register the same pipe for multiple signals, there's `try_clone`
+/// method on many unix socket primitives.
+pub fn register_handle<P>(signal: c_int, pipe: P) -> Result<SigId, Error>
+where
+    P: AsRawHandle + Send + Sync + 'static,
+{
+    let action = move || wake_handle(pipe.as_raw_handle());
+    unsafe { ::register(signal, action) }
+}
+
+/// Registers a write to a self-pipe whenever there's the signal.
+///
+/// The ownership of pipe is taken and will be closed whenever the created action is unregistered.
+///
+/// Note that if you want to register the same pipe for multiple signals, there's `try_clone`
+/// method on many unix socket primitives.
+pub fn register_socket<P>(signal: c_int, pipe: P) -> Result<SigId, Error>
+where
+    P: AsRawSocket + Send + Sync + 'static,
+{
+    let action = move || wake_socket(pipe.as_raw_socket());
+    unsafe { ::register(signal, action) }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::fs::File;
+    use std::io::prelude::*;
+    use std::ptr;
+    use winapi::um::namedpipeapi::CreatePipe;
+    use winapi::um::winnt::HANDLE;
+
+    #[test]
+    fn pipe_notify() -> Result<(), ::std::io::Error> {
+        let (mut read, write) = {
+            let mut read: HANDLE = ptr::null_mut();
+            let mut write: HANDLE = ptr::null_mut();
+            let success = unsafe { CreatePipe(&mut read, &mut write, ptr::null_mut(), 1024) } != 0;
+            assert!(success);
+            let read = unsafe { File::from_raw_handle(read as RawHandle) };
+            let write = unsafe { File::from_raw_handle(write as RawHandle) };
+            (read, write)
+        };
```

Comment:
These blocks are usually introduced to appease the borrow checker. But I don't see why it is needed here. Can you point me to what needs to go out of scope here?

Reply:
Block here was for aligning with this line:

```rust
let (read, write) = UnixStream::pair();
```